### PR TITLE
provider config select

### PR DESCRIPTION
**Pull request by psantos1113**
1. Adds the provider config select field to the distribution UI.  
2. Fixes the backend code to be able to send messages to the channels. 
3. Fixes the email address lookup to use the provider config name.

### CODE REVIEW

File: alert-common/src/main/java/com/synopsys/integration/alert/common/descriptor/config/ui/ProviderDistributionUIConfig.java [hunk 1]
```diff
@@ -75,12 +76,9 @@ public ProviderDistributionUIConfig(String label, String urlName, ProviderConten
 
     @Override
     public List<ConfigField> createFields() {
-        // TODO extract label and description
-        // FIXME add endpoint
-        ConfigField providerConfigName = new EndpointSelectField(ProviderGlobalUIConfig.KEY_PROVIDER_CONFIG_NAME, "Provider Config Name", "The name of the provider configuration to use for this distribution job.")
-                                             .applyClearable(false)
-                                             .applyRequired(true);
-
+        ConfigField providerConfigField = new EndpointSelectField(ProviderDescriptor.KEY_PROVIDER_CONFIG_NAME, ProviderDescriptor.LABEL_PROVIDER_CONFIG_NAME, ProviderDescriptor.DESCRIPTION_PROVIDER_CONFIG_NAME)
```
Comment (gkillough) [inline alert-common/src/main/java/com/synopsys/integration/alert/common/descriptor/config/ui/ProviderDistributionUIConfig.java]: `providerConfigNameField`

File: src/main/java/com/synopsys/integration/alert/provider/blackduck/actions/BlackDuckGlobalTestAction.java [hunk 2]
```diff
@@ -63,7 +64,13 @@ public MessageResult testConfig(String configId, String description, FieldAccess
         String apiToken = fieldAccessor.getStringOrEmpty(BlackDuckDescriptor.KEY_BLACKDUCK_API_KEY);
         String url = fieldAccessor.getStringOrEmpty(BlackDuckDescriptor.KEY_BLACKDUCK_URL);
         String timeout = fieldAccessor.getStringOrEmpty(BlackDuckDescriptor.KEY_BLACKDUCK_TIMEOUT);
-        BlackDuckProperties blackDuckProperties = blackDuckPropertiesFactory.createProperties(Long.valueOf(configId), fieldAccessor);
+        Long parsedConfigurationId = ProviderProperties.UNKNOWN_CONFIG_ID;
+
+        if (StringUtils.isNotBlank(configId)) {
```
Comment (gkillough) [inline src/main/java/com/synopsys/integration/alert/provider/blackduck/actions/BlackDuckGlobalTestAction.java]: Should we validate if configId is a number?

File: alert-common/src/main/java/com/synopsys/integration/alert/common/descriptor/config/ui/ProviderConfigSelectCustomEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        if (descriptorKey.isPresent()) {
+            List<LabelValueSelectOption> options = new LinkedList<>();
+            List<ConfigurationModel> configurationModels = configurationAccessor.getConfigurationByDescriptorKeyAndContext(descriptorKey.get(), ConfigContextEnum.GLOBAL);
+            for (ConfigurationModel configurationModel : configurationModels) {
+                FieldAccessor accessor = new FieldAccessor(configurationModel.getCopyOfKeyToFieldMap());
+                Optional<String> configName = accessor.getString(ProviderDescriptor.KEY_PROVIDER_CONFIG_NAME);
+                if (configName.isPresent()) {
+                    options.add(new LabelValueSelectOption(configName.get()));
+                }
+            }
```
Comment (bamandel) [inline alert-common/src/main/java/com/synopsys/integration/alert/common/descriptor/config/ui/ProviderConfigSelectCustomEndpoint.java]: Could this be converted to a stream?

File: alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultConfigurationAccessor.java [hunk 4]
```diff
@@ -167,15 +167,22 @@ public void deleteJob(UUID jobId) throws AlertDatabaseConstraintException {
...
+                                           .map(FieldValueEntity::getConfigId)
+                                           .collect(Collectors.toList());
+        if (!providerConfigIds.isEmpty()) {
+            for (Long configId : providerConfigIds) {
+                Optional<ConfigurationModel> configurationModel = getConfigurationById(configId);
+                Optional<ConfigurationModel> globalModel = configurationModel
+                                                               .filter(model -> model.getDescriptorContext() == ConfigContextEnum.GLOBAL);
+                if (globalModel.isPresent()) {
+                    return globalModel;
+                }
```
Comment (bamandel) [inline alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultConfigurationAccessor.java]: These lines can all be collapsed into something similar to
`getConfigurationById(configId).filter(model -> model.getDescriptorContext() == ConfigContextEnum.GLOBAL);`
Although exception handling could cause problems here.

File: src/main/java/com/synopsys/integration/alert/provider/blackduck/actions/BlackDuckGlobalTestAction.java (hunk 2 above)
Comment (bamandel) [reply src/main/java/com/synopsys/integration/alert/provider/blackduck/actions/BlackDuckGlobalTestAction.java]: If a configId isn't a number then something is very wrong. I feel it should break in that case.

Comment (gkillough) [reply src/main/java/com/synopsys/integration/alert/provider/blackduck/actions/BlackDuckGlobalTestAction.java]: If someone is using REST calls, configId may not be a number.

File: alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultConfigurationAccessor.java [hunk 5]
```diff
@@ -167,15 +167,22 @@ public void deleteJob(UUID jobId) throws AlertDatabaseConstraintException {
...
+                                           .stream()
+                                           .map(FieldValueEntity::getConfigId)
+                                           .collect(Collectors.toList());
+        if (!providerConfigIds.isEmpty()) {
+            for (Long configId : providerConfigIds) {
+                Optional<ConfigurationModel> configurationModel = getConfigurationById(configId);
+                Optional<ConfigurationModel> globalModel = configurationModel
+                                                               .filter(model -> model.getDescriptorContext() == ConfigContextEnum.GLOBAL);
+                if (globalModel.isPresent()) {
+                    return globalModel;
```
Comment (jamesrichard91) [inline alert-database/src/main/java/com/synopsys/integration/alert/database/api/DefaultConfigurationAccessor.java]: Why are we not just returning the Optional globalModel? What is the purpose of the isPresent check?

File: alert-common/src/main/java/com/synopsys/integration/alert/common/descriptor/config/ui/ProviderDistributionUIConfig.java (hunk 1 above)
Comment (psantos1113) [reply alert-common/src/main/java/com/synopsys/integration/alert/common/descriptor/config/ui/ProviderDistributionUIConfig.java]: done

**Pull request merged**